core/services/ocr2/plugins/s4/integration_test.go

```diff
@@ -0,0 +1,406 @@
+	assert.Empty(t, rows)
+}
+
+func TestS4Integration_HappyDON(t *testing.T) {
+	don := newDON(t, 4, createPluginConfig(100))
+	ctx := testutils.Context(t)
+
+	// injecting new records
+	rows := generateTestOrmRows(t, 10, time.Minute)
+	for _, row := range rows {
+		err := don.orms[0].Update(row, pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+	}
+	originSnapshot, err := don.orms[0].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+	require.NoError(t, err)
+
+	// S4 to propagate all records in one OCR round
+	errors := don.executeOCR(ctx, 1)
+	checkNoErrors(t, errors)
+
+	for i := 0; i < don.size; i++ {
+		snapshot, err := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+		equal := compareSnapshots(originSnapshot, snapshot)
+		assert.True(t, equal, "oracle %d", i)
+		checkNoUnconfirmedRows(ctx, t, don.orms[i], 10)
+	}
+}
+
+func TestS4Integration_HappyDON_4X(t *testing.T) {
+	don := newDON(t, 4, createPluginConfig(100))
+	ctx := testutils.Context(t)
+
+	// injecting new records to all nodes
+	for o := 0; o < don.size; o++ {
+		rows := generateTestOrmRows(t, 10, time.Minute)
+		for _, row := range rows {
+			err := don.orms[o].Update(row, pg.WithParentCtx(ctx))
+			require.NoError(t, err)
+		}
+	}
+
+	// S4 to propagate all records in one OCR round
+	errors := don.executeOCR(ctx, 1)
+	checkNoErrors(t, errors)
+
+	firstSnapshot, err := don.orms[0].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+	require.NoError(t, err)
+
+	for i := 1; i < don.size; i++ {
+		snapshot, err := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+		equal := compareSnapshots(firstSnapshot, snapshot)
+		assert.True(t, equal, "oracle %d", i)
+		checkNoUnconfirmedRows(ctx, t, don.orms[i], 10)
+	}
+}
+
+func TestS4Integration_WrongSignature(t *testing.T) {
+	don := newDON(t, 4, createPluginConfig(100))
+	ctx := testutils.Context(t)
+
+	// injecting new records
+	rows := generateTestOrmRows(t, 10, time.Minute)
+	rows[0].Signature = rows[1].Signature
+	for _, row := range rows {
+		err := don.orms[0].Update(row, pg.WithParentCtx(ctx))
+		require.NoError(t, err)
+	}
+	originSnapshot, err := don.orms[0].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+	require.NoError(t, err)
+	originSnapshot = filter(originSnapshot, func(row *s4_svc.SnapshotRow) bool {
+		return row.Address.Cmp(rows[0].Address) != 0 || row.SlotId != rows[0].SlotId
+	})
+	require.Len(t, originSnapshot, len(rows)-1)
+
+	// S4 to propagate valid records in one OCR round
+	errors := don.executeOCR(ctx, 1)
+	checkNoErrors(t, errors)
+
+	for i := 1; i < don.size; i++ {
+		snapshot, err2 := don.orms[i].GetSnapshot(s4_svc.NewFullAddressRange(), pg.WithParentCtx(ctx))
+		require.NoError(t, err2)
+		equal := compareSnapshots(originSnapshot, snapshot)
+		assert.True(t, equal, "oracle %d", i)
+	}
+
+	// record with a wrong signature must remain unconfirmed
```

Comment:
Maybe check if it doesn't exist on other nodes?

Reply:
We have already verified this above, when comparing with `compareSnapshots` against the first node snapshot that excluded the bad entry. That `compareSnapshots` functions checks for equal length and content.